internal/migrate/versions.go

```diff
@@ -0,0 +1,209 @@
+
+2. Undoing migrations.
+
+	Intentionally left out, because we do not gain that much from currently.
+
+3. Snapshotting the whole state.
+
+	This probably should be done by the user of this library, when there's disk-space available.
+
+4. Figuring out what the exact executed steps are.
+*/
+
+// Migration describes a migration steps
+type Migration struct {
+	Table    string
+	OnCreate Action
+	Steps    []*Step
+}
+
+// Step describes a single step in migration.
+type Step struct {
+	Description string
+	Version     int // Versions should start at 1
+	Action      Action
+}
+
+// Action is something that needs to be done
+type Action interface {
+	Run(log *zap.Logger, db DB, tx *sql.Tx) error
+}
+
+// ValidTableName checks whether the specified table name is valid
+func (migration *Migration) ValidTableName() error {
+	matched, err := regexp.MatchString(`^[a-z_]+$`, migration.Table)
+	if !matched || err != nil {
+		return Error.New("invalid table name: %v", migration.Table)
+	}
+	return nil
+}
+
+// Run runs the migration steps
+func (migration *Migration) Run(log *zap.Logger, db DB) error {
+	err := migration.ValidTableName()
+	if err != nil {
+		return err
+	}
+
+	err = migration.createVersionTable(log, db)
+	if err != nil {
+		return Error.New("creating version table failed: %v", err)
+	}
+
+	version, err := migration.getLatestVersion(log, db)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	if version < 0 {
+		log.Info("Latest Version", zap.Int("version", version))
+	} else {
+		log.Info("No Version")
+	}
+
+	if version < 0 && migration.OnCreate != nil {
+		log := log.Named("init")
+		tx, err := db.Begin()
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		err = migration.OnCreate.Run(log, db, tx)
+		if err != nil {
+			return Error.Wrap(errs.Combine(err, tx.Rollback()))
+		}
+
+		err = migration.addVersion(tx, db, 0)
+		if err != nil {
+			return Error.Wrap(errs.Combine(err, tx.Rollback()))
+		}
+
+		if err := tx.Commit(); err != nil {
+			return Error.Wrap(err)
+		}
+	}
+
+	for _, step := range migration.Steps {
+		if step.Version <= version {
+			continue
+		}
+
+		log := log.Named(strconv.Itoa(step.Version))
+		log.Info(step.Description)
+
+		tx, err := db.Begin()
+		if err != nil {
+			return Error.Wrap(err)
+		}
+
+		err = step.Action.Run(log, db, tx)
+		if err != nil {
+			return Error.Wrap(errs.Combine(err, tx.Rollback()))
+		}
+
+		err = migration.addVersion(tx, db, step.Version)
+		if err != nil {
+			return Error.Wrap(errs.Combine(err, tx.Rollback()))
+		}
+
+		if err := tx.Commit(); err != nil {
+			return Error.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+// createVersionTable creates a new version tabe
+func (migration *Migration) createVersionTable(log *zap.Logger, db DB) error {
+	tx, err := db.Begin()
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	_, err = tx.Exec(db.Rebind(`CREATE TABLE IF NOT EXISTS ` + migration.Table + ` (version int, commited_at text)`))
```

Comment:
why do we use multiple migration tables rather than one w/ an additional 'name' column?

Reply:
I wasn't sure whether we want to hardcode the migration table -- single binary should probably still use a single table.

Reply:
why do we store commited_at as a string?

Reply:
It's mainly for debugging later and STRING is supported on postgres and sqlite in the same way. Mainly avoids some complexity with handling different drivers.

Reply:
Some clarification (by naming or by code comments) around when to use a different 'Migration.Table' would be a good idea.  Different binaries might require the same set of database updates.  Different parts of an application might be entire atomic/unrelated.  The intent is not clear.